### DISCUSSION TITLE
Personalize the TOC based on document category and frontmatter

### DISCUSSION THREAD

**Discussion by LunaticMuch** [facebook/docusaurus#7850] [Q&A] [answered]
I have a case where, for part of the documentation, I need to add some _metadata_ to my page.
Visually, I think the best way is to have a box, above the TOC. 

Pages that requires it can just have, in the frontmatter, something like:

```yaml
displayMetadata: true
```

and then I can have an object like

```yaml
metadata: []
```

Questions are:
1. shall I eject the TOC (which is not safe) or is there another safer alternative?
2. can I read frontmatter info during build time to populate the content?

**Comment by slorber** [comment, marked as answer]
It looks similar to what @lanegoolsby is trying to do, see my answer here: https://github.com/facebook/docusaurus/pull/7644#issuecomment-1209058641

> shall I eject the TOC (which is not safe) or is there another safer alternative?

It depends

The first question to ask yourself is: do you want to customize all Docusaurus TOCs (desktop? mobile?) at once, or just a specific case (desktop only? docs only?)

> can I read frontmatter info during build time to populate the content?

I'm not sure what you mean by that 🤷‍♂️ But:
- For docs, you can use `useDoc().frontMatter.displayMetadata`
- For blog posts, you can use `useBlogPost().frontMatter.displayMetadata`

---

Don't be too afraid by "internal", "unsafe" etc, those are too scary but what we mean in the end is just that we may have to do breaking changes on those apis someday, not that you shouldn't use those. If you only use these on your site's code, then you'd be able to adjust your code once we do breaking changes. It's very unlikely that we'll completely remove the TOC component or the 2 hooks above.

**Reply by LunaticMuch** [reply]
I actually did something that works. 
1. Swizzled the TOC
2. used `import useFrontMatter from "@theme/useFrontMatter";`
3. extracted what I needed from the frontmatter with `const { metadata } = useFrontMatter();`

**Reply by slorber** [reply]
Great 👍 

`useFrontMatter()` is not a core api (but may be someday), I guess you are using https://github.com/roydukkey/docusaurus-theme-frontmatter ?

**Comment by lanegoolsby** [comment]
I can swizzle the TOC component fine, that works great. Where I'm stuck is in trying to convert the swizzled component into a custom plugin.

**Comment by lanegoolsby** [comment]
I have a [working example](https://github.com/facebook/docusaurus/discussions/7900#discussioncomment-3360504) using the `useDoc` hook Sebastian mentioned above. 

My example is using a custom plugin, so a bunch of that is unnecessary if you can get away with swizzling. In my case I need the custom code to be reusable across multiple sites so needing to create the plugin complicates things a little.
